src/transforms/kubernetes/watch_client.rs

```diff
@@ -0,0 +1,469 @@
+use crate::dns::Resolver;
+use futures::{future::Future, stream::Stream};
+use http::{header, status::StatusCode, uri, Request, Uri};
+use hyper::{client::HttpConnector, Body, Chunk};
+use hyper_openssl::HttpsConnector;
+use k8s_openapi::{
+    api::core::v1::{Pod, WatchPodForAllNamespacesResponse},
+    apimachinery::pkg::apis::meta::v1::WatchEvent,
+    RequestError, Response, ResponseError, WatchOptional,
+};
+use openssl::ssl::{SslConnector, SslConnectorBuilder, SslMethod};
+use snafu::{ResultExt, Snafu};
+
+/// Config which could be loaded from kubeconfig or local kubernetes cluster.
+/// Used to build WatchClient which is in turn used to build Stream of metadata.
+pub struct ClientConfig {
+    resolver: Resolver,
+    token: Option<String>,
+    server: Uri,
+    ssl: Option<SslConnectorBuilder>,
+    node_name: Option<String>,
+}
+
+impl ClientConfig {
+    pub fn build(mut self) -> Result<WatchClient, BuildError> {
+        let mut http = HttpConnector::new_with_resolver(self.resolver.clone());
+        http.enforce_http(false);
+
+        let ssl = if let Some(ssl) = self.ssl.take() {
+            ssl
+        } else {
+            SslConnector::builder(SslMethod::tls()).context(OpenSslError)?
+        };
+
+        let https = HttpsConnector::with_connector(http, ssl).context(OpenSslError)?;
+
+        let client = hyper::Client::builder().build(https);
+
+        let client = WatchClient {
+            client,
+            config: self,
+        };
+
+        Ok(client)
+    }
+
+    // Builds request to watch Pod data.
+    fn build_request(&self, version: Option<Version>) -> Result<Request<Body>, BuildError> {
+        // Prepare request
+        let field_selector = self
+            .node_name
+            .as_ref()
+            .map(|node_name| format!("spec.nodeName={}", node_name));
+        let (mut request, _) = Pod::watch_pod_for_all_namespaces(WatchOptional {
+            resource_version: version.as_ref().map(|v| v.0.as_str()),
+            field_selector: field_selector.as_ref().map(|selector| selector.as_str()),
+            ..WatchOptional::default()
+        })
+        .context(K8SOpenapiError)?;
+
+        self.authorize(&mut request)?;
+        self.fill_uri(&mut request)?;
+
+        let (parts, body) = request.into_parts();
+        Ok(Request::from_parts(parts, body.into()))
+    }
+
+    fn authorize(&self, request: &mut Request<Vec<u8>>) -> Result<(), BuildError> {
```

Comment:
minor nit: in other parts of our codebase we tend to not use extra fn like this but would prefer to just have all these within the build_request fn.